### ISSUE TITLE
Load calculation for gas: specification

### ISSUE THREAD

**Issue by ChaelKruip** [quintel/etmoses#861]
Contrary to the load calculation for electricity, there is no flexible network topology but rather a static 'stack' of buffers which communicate with each other. The gas topology can be visualised as follows:
![20141209_topology_interface_v1_key](https://cloud.githubusercontent.com/assets/1303760/13924571/2a481620-ef85-11e5-855d-befa7bd10f65.png)
The levels, and the connectors have the same structure and functionality at all levels. The attributes included in the above image are variable and are directly determined by the choices the user makes in the [asset page](https://github.com/quintel/etmoses/issues/850#issuecomment-196768311).

Technologies will in general be connected to the lowest level but can also be connected higher up:
![20141209_topology_interface_v1_key](https://cloud.githubusercontent.com/assets/1303760/13924779/e0f6dee2-ef85-11e5-9e40-bdd23ebcd72b.png)

## Relation between energy pressure and volume

For each level, the pressure can be derived from the energy and the volume by using:
```
pressure = constant * energy / volume.
```

Where the `constant` equals 36 [bar * m^3 / kWh].

## Calculation steps

The calculation steps will likely be something like (for each time-step):

1. start at lowest gas level and add/subtract all load of connected technologies.
2. update pressure level by adding the effective amount of gas (kWh) to the current amount of gas and using the above formula. 
3. check if the pressure level is outside of the range which the level can accommodate.
  * IF the pressure is higher than the upper limit of the range, export gas to the higher level by compressing it (with a given efficiency)
  * IF the pressure is lower than the lower limit, import from the higher level by decompressing it (with a given efficiency)
4. keep track of the sign of the flow (did this level push gas up, or did it pull gas down?), the size of the flow and the losses
5. go to the next level

**Comment by ChaelKruip** [comment]
> The attributes included in the above image are variable and are directly determined by the choices the user makes in the asset page.

To elaborate:

* The (storage) volume of a pressure level is related to the length and diameter of pipes on that level by:
```
volume [m^3] = PI * (pipe_diameter [m])^2 * pipe_length [m]
```
for all pipes.

* The capacity (both 'up' and 'down') is determined by the joint capacity of all 'connectors' to the next level:
```
capacity [kW] = connector_capacity_1 + connector_capacity_2 + connector_capacity_3 + ...
```

**Comment by ChaelKruip** [comment]
NOTE: the lowest pressure level (0.125 bar) is *actually* 1.125 bar. The 0.125 used in practice and speech is actually an **overpressure** added to the standard atmospheric pressure of 1.0 bar. For higher pressure levels, such as 4 bar, I'm not sure if that is an overpressure as well.

@Sanne512 @JeroenvdLogt do you know this?

**Comment by Sanne512** [comment]
That's totally correct, every value mentioned is an overpressure.

**Comment by ChaelKruip** [comment]
After discussion with experts within Alliander, @Sanne512 and co have come to the conclusion that modelling of conventional buffering is not relevant (the potential is negligible). This simplifies the picture described above in the following sense:

* The volume of pipes is no longer relevant
* The pressure and pressure ranges are not needed
* For every level, conservation of energy is the only thing that determines the flow

Note that it is still interesting to show the load [kW], yearly feed-in/de-central production [kWh], yearly consumption/extraction [kWh] and losses per level. The losses are still determined by the efficiencies of the 'connectors'.

**Comment by ChaelKruip** [comment]
Including @dennisschoenmakers @grdw @Sanne512

**Comment by antw** [comment]
> The pressure and pressure ranges are not needed
> For every level, conservation of energy is the only thing that determines the flow

This being the case, what is the significance of `capacity` on connections? With pressure modelled, we could have constrained the flow of gas by capacity, and then shown times of the year when a layer was under- or over-pressure.

Is the capacity still a hard constraint limiting the flow from one level to another, or a soft constraint which may be exceeded but results in the connection being shown in red (like the electricity network)?

**Comment by ChaelKruip** [comment]
> Is the capacity still a hard constraint limiting the flow from one level to another, or a soft constraint which may be exceeded but results in the connection being shown in red (like the electricity network)?

I think the capacity is a hard constraint both up and down (they are not equal). Downwards the gas is expanding and the maximum rate at which this happens is given by how big the 'hole' is and how fast the gas can be heated. Upwards (from low to high pressure), the capacity is constrained (hard) by what the compressor can do. 

So, I think instead of showing 'congestions', it might be more useful to inform the user of how much of the gas-demand is not satisfied (for downward flow).
For upward flow (feeding in hydrogen from P2G installations) I think we should limit the feeding-in when the (upward) capacity is constrained. So the owner of the P2G installation is penalised by not being able to convert all his excess PV production into gas.

@Sanne512 do you agree with this assessment?

**Comment by Sanne512** [comment]
Capacitiy is indeed a hard constraint both up and down. The gas moves downward due to a pressure difference caused by gas use. The biggest restraint for the maximum rate is indeed the diameter of the pipe. Upwards the capacity is indeed (hard) constrained by the compressor. 

About what to show: the gas-demand is being met almost always, due to the many "vermazingen". More specific: the gas infrastucture is developed with so many vermazingen to keep pressure from the gas in the pipes in all times. So I don't think how much the gas-demand is not satisfied is interesting to show.
For upward flow I agree, that the feeding-in should be limited when the capacatiy is constrained. What happens in practice is that when there is no capacity for the fed-in gas it is being flared. 

Besides the consequences and possibilities for feeding-in the costs are the most interesting to show, because they make it possible to compare the gas infrastructure with heat nets or going all-electric with heat pumps.

**Comment by antw** [comment]
@ChaelKruip and @Sanne512: Thank you for your input. I'm going to think about what you've written, and how it impacts the gas model, and will get back to you shortly with a more detailed reply.

In the meantime, I do have one further question: how far do the feed-in constraints apply? To the level at which the technology is attached, or all the way to the highest level?

For example:

```
+---------+
| Level 1 |
+---------+
     |        capacity: 5kW
+---------+
| Level 2 |
+---------+
     |        capacity: 8kW
+---------+
| Level 3 |
+---------+
     |        capacity: infinite
+---------+
|  Tech.  |   production: 10kW
+---------+
```

Does the technology...

1. Produce 10kW, since it can supply that to Level 3, or
2. Produce 8kW, since Level 3 – the level to which it is attached – can supply that to Level 2, or
3. Only produce 5kW as the connection between Level 1 and Level 2 is limiting.

If (1) or (2), energy goes missing and is no longer conserved (the tech produces 10kW, but only 8kW reaches Level 2, and only 5kW to Level 1). And if it is (2) or (3), then I have more thinking to do. :anguished:

**Comment by ChaelKruip** [comment]
> And if it is (2) or (3), then I have more thinking to do. :anguished:

I would say (3). But the way in which this would happen is still bottom-up in my mind. So, the bottom level has no immediate knowledge of the top level constraints.

What I think would happen (please correct me if I'm wrong @Sanne512):

* 10 kW is fed in to level 3
* 8 kW is fed into level 2, 2 kW is 'flared' at the intersection between level 2 and level 3
* 5 kW is fed into level 1, 3 kW is flared

So, every step upwards can be made with only information of the previous level...
Hope that does not require too much re-thinking!

**Comment by ChaelKruip** [comment]
The above implies that it's useful to keep track of how much is flared at each intersection...

**Comment by antw** [comment]
> 2 kW is 'flared' at the intersection [...] Hope that does not require too much re-thinking!

Thanks! That is good news. :tada:

**Comment by antw** [comment]
... although that does assume that there is an inverse for flaring when there is too much **consumption**. If you take [the above example](https://github.com/quintel/etmoses/issues/861#issuecomment-205807986) and swap "production" for "consumption", presumably the technology consumes 10kW, but only 8kW appears on Level 2, and 5kW on Level 1.

This means that the connection between Level 2 and 3 needs to record a deficit of 2kW, and Level 1 and 2 a deficit of 5kW. Does this still make sense?

**Comment by ChaelKruip** [comment]
> Does this still make sense?

That makes sense to me. In real life, the infrastructure for consumption always will have sufficient capacity. In the case that the user makes a LES with capacities which are limiting for consumption, I think it would be best to keep track (and display!) deficits at the various connections. Both flaring and deficits would mean that extra investments need to be made.

**Comment by antw** [comment]
Is the capacity of a connection an *input* capacity or an *output* capacity? The order in which we apply the constraints (capacity and efficiency) affect the energy coming out the other end, and the amount of loss and flaring.

For example:

```
Input Capacity                      |    Output Capacity
                                    |
+---+                               |    +---+
|   |  Layer a: 4.0kW production    |    |   |  Layer a: 4.0kW production
+---+                               |    +---+
  |                                 |      |
  v                                 |      v
 ( )   capacity: 2.0kW              |     ( )   efficiency: 0.8
  |    (2kW sent, 2kW flared)       |      |    (3.2kW sent, 0.8kW lost)
  v                                 |      v
 ( )   efficiency: 0.8              |     ( )   capacity: 2.0kW
  |    (1.6kW sent, 0.4kW lost)     |      |    (2kW sent, 1.2kW flared)
  v                                 |      v
+---+                               |    +---+
|   |  Layer b: 1.6kW received      |    |   |  Layer b: 2kW received
+---+                               |    +---+
```

**Comment by ChaelKruip** [comment]
> Is the capacity of a connection an input capacity or an output capacity?

Output. The flaring is applied only after all losses have been accounted for I think.

Note, however, that the capacity and loss depend on the *direction* the gas is flowing. When flowing from high to low pressure, capacity is very large and losses small. Upwards the capacity is limited and losses substantial.

**Comment by antw** [comment]
Thanks again @ChaelKruip. That is how I implemented the connections already, so there is nothing to change (although changing it would have been trivial).

> Note, however, that the capacity and loss depend on the direction the gas is flowing.

That's fine; I've modelled the connections as having different capacities and efficiencies depending on the direction.

**Comment by ChaelKruip** [comment]
Can we close this @antw?
